[containerapp] Update builder image and patch logic

cormacpayne opened the pull request:
This PR aims to update the following for the `containerapp` extension:
- Moves from a Mariner-based builder to a set of Debian-based builders created on top of the Oryx build system
- Updates the patching logic to reflect the new builders, and more specifically, the new runtime images that images will be based on and patched from
- Tests expanded to verify that both builders are targeted correctly based on the version supported by the provided app
- Tests expanded to verify that patching works for new platforms supported by the builders
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### Related command
<!--- Please provide the related command with az {command} if you can, so that we can quickly route to the related person to review. --->

az containerapp up
az containerapp patch list
az containerapp patch apply


### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally? (`pip install wheel==0.30.0` required)

For new extensions:

~- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).~


### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

yonzhan commented:
<!-- add-pr-comment:add-pr-comment -->

containerapp

daniv-msft commented on src/containerapp/azext_containerapp/_up_utils.py:
```diff
@@ -478,61 +479,74 @@ def build_container_from_source_with_buildpack(self, image_name, source):  # pyl
-                    raise CLIError("Non-zero exit code returned from 'pack build'; please check the above output for more details.")
+                        # Check if the application is targeting a platform that's found in the current builder
+                        if not is_non_supported_platform and "No buildpack groups passed detection" in stdout_line:
+                            is_non_supported_platform = True
```
Could you please confirm: should we leave early the while loop when this happens? Any reasons to stick with it or can we directly move to the next builder? Same comment below

cormacpayne replied:
I'm going to keep this as-is for now -- typically, these messages are going to be at the end of the process anyway, so there wouldn't be much more to process before the flow moves on to the next builder, but more importantly, I'm not entirely sure what the ramifications of exiting the process early are without first calling `process.communicate()`. I can test this out to be sure that it won't affect the next builder call, but I also don't think it would make a difference performance-wise if we kept it as-is.

daniv-msft commented on src/containerapp/azext_containerapp/_up_utils.py:
```diff
@@ -478,61 +479,74 @@ def build_container_from_source_with_buildpack(self, image_name, source):  # pyl
-                if is_non_supported_platform:
-                    raise ValidationError("Current buildpacks do not support the platform targeted in the provided source code.")
+                        stdout_line = str(line.strip(), 'utf-8')
+                        logger.warning(stdout_line)
```
I'm wondering how valuables are the logs before we've validated for sure that the current builder is the right one. I'm not feeling strongly, but would we lose much if we only started showing the logs once we know the builder will build?

cormacpayne replied:
Discussed offline -- we will capture the stdout in a separate variable that we'll flush _only_ when the builder was successful in building the provided application source.

daniv-msft commented on src/containerapp/azext_containerapp/_up_utils.py:
```diff
@@ -478,61 +479,74 @@ def build_container_from_source_with_buildpack(self, image_name, source):  # pyl
-            raise ex
-        except Exception as ex:
-            raise CLIError(f"Unable to run 'pack build' command to produce runnable application image: {ex}") from ex
+                        # Check if the application has a version that it can target from the current builder
```
Nit: please consider updating this comment to talk about the runtime images and all. Maybe it's just me, but this comment as of today feels very similar to the other `Check if the application is targeting a platform that's found in the current builder` above

daniv-msft commented on src/containerapp/azext_containerapp/_utils.py:
```diff
@@ -1970,136 +1973,161 @@ def get_pack_exec_path():
+                            with open(exec_zip_path, "wb") as f:
                                 f.write(tar.extractfile(tar_info).read())
 
+        # Rename the executable to include the CLI version (to ensure future CLI versions are installed and consumed)
```
Please consider updating the comment to make it clear that here, "CLI" is the "pack CLI", not the "ACA CLI"?

daniv-msft commented on src/containerapp/azext_containerapp/custom.py:
```diff
@@ -5149,7 +5135,7 @@ def patch_apply_handle_input(cmd, patch_check_list, method, pack_exec_path):
     try:
         logger.warning("Applying patch for container app: " + container_app_name + " container: " + container_name)
-        subprocess.run(f"{pack_exec_path} rebase -q {target_image_name} --run-image {new_run_image}", shell=True, check=True)
+        subprocess.run(f"{pack_exec_path} rebase -q {target_image_name} --run-image {new_run_image} --force", shell=True, check=True)
```
Could you please clarify why we need `--force` now? What has changed?

cormacpayne replied:
I'll let @harryli0108 provide additional context, but AFAIK, the `--force` flag is now needed for the `pack rebase` command base on some of the latest changes made to their lifecycle to support our specific scenario.

harryli0108 replied:
The `--force` is introduced because we couldn't include new releases into the image metadata. Pack rebase checks if the new run image is included in the existing image's metadata. This requires us to force rebasing the image. Otherwise, the rebase will fail with the following error: 
```
failed to rebase: rebase app image: new base image 'mcr.microsoft.com/oryx/builder:run-dotnet-aspnet-7.0.7-cbl-mariner2.0' not found in existing run image metadata: {"image":"mcr.microsoft.com/oryx/builder:run-dotnet-aspnet-7.0-cbl-mariner2.0"}

daniv-msft commented:
@Greedygre, @howang-ms Hello! We would really appreciate if you could review this PR as I believe we need a sign-off from one of you to merge it (could you please confirm?).

cormacpayne commented:
@Greedygre thanks for the comment -- I pulled in the latest changes from `main`, resolved the feedback and recorded all tests necessary for our flow.

Please feel free to let me know if there's anything additional needed for this PR.

Greedygre commented on src/containerapp/HISTORY.rst:
```diff
@@ -13,6 +13,7 @@ upcoming
 * 'az containerapp job create/update': hide environment variables, scale rules metadata, eventTriggerConfig for job
 * [Breaking Change] 'az containerapp env create': update the default value of --enable-workload-profiles to `True`
 * 'az containerapp compose create': fix containerapp invalid memory resource
+* 'az containerapp up': update builder image used when --source is provided with no Dockerfile to support building applications targeting a wider range of platform versions
```
Please move to the `upcoming `part.

Greedygre commented:
Please get approval from your team who could be responsible for your business logic.

daniv-msft commented:
> Please get approval from your team who could be responsible for your business logic.

@Greedygre I confirm I'm in the same team as @cormacpayne and the changes look good to me. Thanks

Greedygre commented:
Hi @cormacpayne 

In 0.3.42, we updated the cluster for environment from V1 to V2. 
I notice in this recording file, the environment is still V1 cluster. 
Please make sure rerun the test Pass lively with V2 cluster. 

src/containerapp/azext_containerapp/tests/latest/recordings/test_containerapp_patch_list_and_apply_with_node18_e2e.yaml

```
- request:
    body: '{"location": "eastus", "tags": null, "properties": {"daprAIInstrumentationKey":
      null, "vnetConfiguration": null, "appLogsConfiguration": {"destination": "log-analytics",
      "logAnalyticsConfiguration": {"customerId": "mock",
      "sharedKey": "mock"}},
      "customDomainConfiguration": null, "workloadProfiles": null, "InfrastructureResourceGroup":
      null, "zoneRedundant": false}}'
    headers:
      Accept:
      - '*/*'
      Accept-Encoding:
      - gzip, deflate
      CommandName:
      - containerapp env create
      Connection:
      - keep-alive
      Content-Length:
      - '483'
      Content-Type:
      - application/json
      ParameterSetName:
      - -g -n
      User-Agent:
      - python/3.8.10 (Windows-10-10.0.22621-SP0) AZURECLI/2.53.0
    method: PUT

```

cormacpayne commented:
@Greedygre thanks for the heads up -- I just merged the latest changes from `main`, updated my `containerapp` module to latest and re-recorded all of the tests I added/modified in this PR. Please feel free to let me know if there's anything additional you need from me!

Greedygre commented:
LGTM. @Juliehzl Could you help to review this PR?

Greedygre commented on src/containerapp/azext_containerapp/custom.py:
```diff
@@ -1100,115 +1107,94 @@ def patch_list(cmd, resource_group_name=None, managed_env=None, show_all=False):
-                            newRunImage=None,
-                            id=None,
-                            reason=mcr_check_reason))
+        results.append(_get_patchable_check_result(inspect_result, oryx_run_images))
```
Should we move `_get_patchable_check_result` to `_utils` or `patch_utils` in the feature?

Greedygre commented:
Hi @cormacpayne please fix the conflict in history.rst.

Juliehzl commented on src/containerapp/azext_containerapp/_up_utils.py:
```diff
@@ -464,7 +466,7 @@ def _docker_push_to_container_registry(self, image_name, forced_acr_login=False)
             raise CLIError(f"Unable to run 'docker push' command to push image to the container registry: {ex}") from ex
 
-    def build_container_from_source_with_buildpack(self, image_name, source):  # pylint: disable=too-many-statements
+    def build_container_from_source_with_buildpack(self, image_name, cache_image_name, source):  # pylint: disable=too-many-statements
```
a good practice: add argument at last position to avoid positional argument mismatch

Juliehzl commented on src/containerapp/azext_containerapp/_utils.py:
```diff
@@ -282,136 +285,161 @@ def get_pack_exec_path():
+                tag_dict[framework][version_key][support] = {os_prop: [tag_obj]}
+            elif os_prop not in tag_dict[framework][version_key][support]:
+                tag_dict[framework][version_key][support][os_prop] = [tag_obj]
             else:
```
will the response always follow the expected format?

cormacpayne replied:
I'm optimistic that the MCR team will do their part to avoid breaking changes within their `v2` API that we're calling here, but since we're not the owners of this registry (or the API), I'm not personally able to make any guarantees that this expected format will _always_ remain the same.

daniv-msft replied:
Yes, I confirm it's fine. This scenario should never happen, and if the MCR team introduces a breaking change that would break this convention, then it would be the kind of scenario that would deserve an exception to be raised. :)

Juliehzl reviewed (approved):
:shipit:

Greedygre commented:
Hi @yanzhudd @wangzelin007 

Could you help to review and merge this PR? As other PR relay on this PR. Hope to get merged for this PR. Thanks.
